[FSDP][optim_state_dict] Enable cpu_offload config for optimzer state_dict

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108434

We had the option but never used cpu_offload as optimizer state_dict offloads the tensors to CPU by default. And this is usually most users want as the tensors are required to be moved to CPU eventually. However, we may want to disable offloading to CPU in some cases, epsecially for the debugging purpose. This PR lets optimizer state_dict read the flag.

Differential Revision: [D48913340](https://our.internmc.facebook.com/intern/diff/D48913340/)

fegin committed:
[FSDP][optim_state_dict] Enable cpu_offload config for optimzer state_dict

We had the option but never used cpu_offload as optimizer state_dict offloads the tensors to CPU by default. And this is usually most users want as the tensors are required to be moved to CPU eventually. However, we may want to disable offloading to CPU in some cases, epsecially for the debugging purpose. This PR lets optimizer state_dict read the flag.

Differential Revision: [D48913340](https://our.internmc.facebook.com/intern/diff/D48913340/)

[ghstack-poisoned]

wz337 commented on test/distributed/fsdp/test_fsdp_optim_state.py:
```diff
@@ -1854,8 +1855,40 @@ def step():
+                    if s.dim() == 0:
+                        continue
+                    for s in state.values():
+                        self.assertTrue(s.is_cuda)
```
Should is_cuda be False since offload_to_cpu=True?

fegin committed:
Update on "[FSDP][optim_state_dict] Enable cpu_offload config for optimzer state_dict"


We had the option but never used cpu_offload as optimizer state_dict offloads the tensors to CPU by default. And this is usually most users want as the tensors are required to be moved to CPU eventually. However, we may want to disable offloading to CPU in some cases, epsecially for the debugging purpose. This PR lets optimizer state_dict read the flag.

Differential Revision: [D48913340](https://our.internmc.facebook.com/intern/diff/D48913340/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Enable cpu_offload config for optimzer state_dict"


We had the option but never used cpu_offload as optimizer state_dict offloads the tensors to CPU by default. And this is usually most users want as the tensors are required to be moved to CPU eventually. However, we may want to disable offloading to CPU in some cases, epsecially for the debugging purpose. This PR lets optimizer state_dict read the flag.

Differential Revision: [D48913340](https://our.internmc.facebook.com/intern/diff/D48913340/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Enable cpu_offload config for optimzer state_dict"


We had the option but never used cpu_offload as optimizer state_dict offloads the tensors to CPU by default. And this is usually most users want as the tensors are required to be moved to CPU eventually. However, we may want to disable offloading to CPU in some cases, epsecially for the debugging purpose. This PR lets optimizer state_dict read the flag.

Differential Revision: [D48913340](https://our.internmc.facebook.com/intern/diff/D48913340/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Enable cpu_offload config for optimzer state_dict"


We had the option but never used cpu_offload as optimizer state_dict offloads the tensors to CPU by default. And this is usually most users want as the tensors are required to be moved to CPU eventually. However, we may want to disable offloading to CPU in some cases, epsecially for the debugging purpose. This PR lets optimizer state_dict read the flag.

Differential Revision: [D48913340](https://our.internmc.facebook.com/intern/diff/D48913340/)

[ghstack-poisoned]

wz337 reviewed (approved):
LGTM!

fegin committed:
Update on "[FSDP][optim_state_dict] Enable cpu_offload config for optimzer state_dict"


We had the option but never used cpu_offload as optimizer state_dict offloads the tensors to CPU by default. And this is usually most users want as the tensors are required to be moved to CPU eventually. However, we may want to disable offloading to CPU in some cases, epsecially for the debugging purpose. This PR lets optimizer state_dict read the flag.

Differential Revision: [D48913340](https://our.internmc.facebook.com/intern/diff/D48913340/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Enable cpu_offload config for optimzer state_dict"


We had the option but never used cpu_offload as optimizer state_dict offloads the tensors to CPU by default. And this is usually most users want as the tensors are required to be moved to CPU eventually. However, we may want to disable offloading to CPU in some cases, epsecially for the debugging purpose. This PR lets optimizer state_dict read the flag.

Differential Revision: [D48913340](https://our.internmc.facebook.com/intern/diff/D48913340/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Enable cpu_offload config for optimzer state_dict"


We had the option but never used cpu_offload as optimizer state_dict offloads the tensors to CPU by default. And this is usually most users want as the tensors are required to be moved to CPU eventually. However, we may want to disable offloading to CPU in some cases, epsecially for the debugging purpose. This PR lets optimizer state_dict read the flag.

Differential Revision: [D48913340](https://our.internmc.facebook.com/intern/diff/D48913340/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Enable cpu_offload config for optimzer state_dict"


We had the option but never used cpu_offload as optimizer state_dict offloads the tensors to CPU by default. And this is usually most users want as the tensors are required to be moved to CPU eventually. However, we may want to disable offloading to CPU in some cases, epsecially for the debugging purpose. This PR lets optimizer state_dict read the flag.

Differential Revision: [D48913340](https://our.internmc.facebook.com/intern/diff/D48913340/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge
